File: maximum-twin-sum-of-a-linked-list/solution2130.py
# https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/

from typing import Optional


class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution2130:

    # T: O(n), S: O(1)
    def pairSum(self, head: Optional[ListNode]) -> int:
        first_half_end = self.find_first_half_end(head)
        second_half_start = self.reverse_list(first_half_end.next)
        max_twin_sum = 0
        curr1, curr2 = head, second_half_start
        while curr2:
            curr_sum = curr1.val + curr2.val
            if curr_sum > max_twin_sum:
                max_twin_sum = curr_sum
            curr1 = curr1.next
            curr2 = curr2.next
        first_half_end.next = self.reverse_list(second_half_start)
        return max_twin_sum

    # T: O(n), S: O(1)
    def find_first_half_end(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    # T: O(n), S: O(1)
    def reverse_list(self, node: ListNode) -> ListNode:
        prev = None
        curr = node
        while curr:
            temp_next = curr.next
            curr.next = prev
            prev = curr
            curr = temp_next
        return prev


def main() -> None:
    sol = Solution2130()
    head1 = ListNode(5, ListNode(4, ListNode(2, ListNode(1))))
    head2 = ListNode(4, ListNode(2, ListNode(2, ListNode(3))))
    head3 = ListNode(1, ListNode(100000))
    print(sol.pairSum(head1))  # 6
    print(sol.pairSum(head2))  # 7
    print(sol.pairSum(head3))  # 100001


if __name__ == "__main__":
    main()

File: maximum-twin-sum-of-a-linked-list/solution2130.go
// https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// T: O(n), S: O(1)
func pairSum(head *ListNode) int {
	firstHalfEnd := findFirstHalfEnd(head)
	secondHalfStart := reverseList(firstHalfEnd.Next)
	maxTwinSum := 0
	curr1, curr2 := head, secondHalfStart
	for curr2 != nil {
		currSum := curr1.Val + curr2.Val
		if currSum > maxTwinSum {
			maxTwinSum = currSum
		}
		curr1 = curr1.Next
		curr2 = curr2.Next
	}
	firstHalfEnd.Next = reverseList(secondHalfStart)
	return maxTwinSum
}

// T: O(n), S: O(1)
func findFirstHalfEnd(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

// T: O(n), S: O(1)
func reverseList(node *ListNode) *ListNode {
	var prev *ListNode
	curr := node
	for curr != nil {
		tempNext := curr.Next
		curr.Next = prev
		prev = curr
		curr = tempNext
	}
	return prev
}

func main() {
	head1 := &ListNode{5, &ListNode{4, &ListNode{2, &ListNode{1, nil}}}}
	head2 := &ListNode{4, &ListNode{2, &ListNode{2, &ListNode{3, nil}}}}
	head3 := &ListNode{1, &ListNode{100000, nil}}
	fmt.Println(pairSum(head1)) // 6
	fmt.Println(pairSum(head2)) // 7
	fmt.Println(pairSum(head3)) // 100001
}
